### ISSUE TITLE
crashing navigation.navigate call because _history undefined

### ISSUE THREAD

**Issue by vongohren** [frontarm/navi#98]
# Issue
What happens is that i ask the user to load a url with a verification token in it.
My `[routeRegistry.login.verifyToken.path]` should pick this up, and it does. It then loads the VerifyToken component in the bottom of this issue.

But when everything goes fine, and it is suppose to call navigate, navigate crashes on this line https://github.com/frontarm/navi/blob/master/packages/navi/src/Navigation.ts#L182
It says _history is undefined.

How come?

# Code setup
I have this route setup for now

```
export default mount({
  '*': map((request, context) => {
    const { authService } = context
    if(_.isEmpty(authService.currentUser)){
      const originalUrl = encodeURIComponent(request.originalUrl)
      return redirect(
        `${routeRegistry.login.path}?redirectTo=${originalUrl}`,
        {exact: false}
      )
    }
    return getAuthenticatedRoutes();
  }),
  [routeRegistry.root.path]: redirect(routeRegistry.login.path),
  [routeRegistry.login.path]: map(async (request, context) => {
    const { authService } = context
    if(!_.isEmpty(authService.currentUser)) {
      const path = decidedRedirectPath(request.params.redirectTo, routeRegistry.dashboard.path)
      return redirect(path)
    } else {
      return getLoginRoute();
    }
  }),
})

const getLoginRoute = () => {
  return mount({
    '/': route({
      title: 'Login',
      view: <SignUp />
    }),
    [routeRegistry.login.verifyToken.path]: route((req,context) => {
      const token = req.params.token
      return {
        title: 'Verifying...',
        view: <TokenVerification token={token} authService={context.authService} />,
      }
    }),
  })
}
```
With this verifyToken component
```
import { useNavigation } from 'react-navi';

export default ({token, authService}) => {
  let navigation = useNavigation();
  useEffect(() => {
    const verifyEmailAsync = async () => {
      const { res } = await verifyEmail(token);
      if(res.status === 200) {
        await authService.login(res.data.token)
        navigation.navigate('/dashboard');
      }
    }
    verifyEmailAsync();
  }, [])

  return (
    <div>Loading..</div>
  );
```

**Comment by jamesknelson** [comment]
Usually, the only way that `this._history` can be undefined is that the `<Router>` component itself has been unmounted.

Could you show me how your `<Router>` component is being defined? Is it at the top level or somewhere lower?

**Comment by vongohren** [comment]
@jamesknelson So I have the router defined at the highest level.

Above this is just the app inject into the DOM.
So dont understand why the router is ejected?
Can it have something to do with how I mount and stuff?
That the current route is not mounted somehow?
Or should it be available anyways?
Have added all my router logic underneath.
It happens under getNonAuthenticatedRoutes, and tokenVerification. We load the page from a link in an email and verify their login that way.

```
const RouterLayer = () => {
  const authService = AuthService.useContainer();
  if(_.isEmpty(authService.currentUser) && authService.loading) authService.login();
  if(authService.loading) return <LoadLayout authService={authService} />

  return (
    <Router routes={routes} context={{authService}}>
      <MainLayout authService={authService}>
        <Suspense fallback={null}>
          <View />
        </Suspense>
      </MainLayout>
    </Router>
  )
}

const ProviderLayer = () => {
  return (
    <AuthService.Provider>
      <RouterLayer />
    </AuthService.Provider>
  )
}

export default ProviderLayer;

export default mount({
  '*': map((request, context) => {
    const { authService } = context
    if(_.isEmpty(authService.currentUser)){
      const originalUrl = encodeURIComponent(request.originalUrl)
      return redirect(
        `${routeRegistry.login.path}?redirectTo=${originalUrl}`,
        {exact: false}
      )
    }
    return getAuthenticatedRoutes();
  }),
  [routeRegistry.root.path]: redirect(routeRegistry.login.path),
  [routeRegistry.login.path]: map(async (request, context) => {
    const { authService } = context
    if(!_.isEmpty(authService.currentUser)) {
      const path = decidedRedirectPath(request.params.redirectTo, routeRegistry.dashboard.path)
      return redirect(path)
    } else {
      return getNonAuthenticatedRoutes();
    }
  }),
  [routeRegistry.invitation.path]: route((request, context)=>{
    const { authService } = context
    const token = request.params.token
    return {
      title: 'Decentralized ID login',
      view: <InvitationVerification authService={authService} token={token} />
    }
  }),
  [routeRegistry.admin.path]: map(async (req, context) => {
    const { authService } = context
    const { currentUser } = authService
    if(!currentUser.isAdmin) {
      Notify.error('You do not have admin rights');
      return redirect(routeRegistry.root.path);
    }
    return route({
      title: 'Diwala admin page',
      view: <Admin />
    })
  }),
})

const getNonAuthenticatedRoutes = () => {
  return mount({
    '/': route({
      title: 'Login to Diwala',
      view: <SignUp />
    }),
    [routeRegistry.login.onboard.path]: route({
      title: 'How decentralized ID works',
      view: <OnboardWizard />
    }),
    [routeRegistry.login.verifyToken.path]: route((request,context) => {
      const token = request.params.token
      return {
        title: 'Verifying...',
        view: <TokenVerification token={token} authService={context.authService} />,
      }
    }),
    [routeRegistry.login.didLogin.path]: route((request, context)=>{
      const { authService } = context
      return {
        title: 'Decentralized ID login',
        view: <DidLogin authService={authService}/>
      }
    })
  })
}

const getCoursesRoute = () => {
  return mount({
    '/': map(async (request,context) => {
      return route({
        title: "Your courses",
        view: <Courses/>,
      })
    }),
    '/:id': map(async (request,context) => {
      return route({
        title: "Specific course",
        view: <Course/>,
      })
    })
  })
}

const decidedRedirectPath = (redirectParams, fallback) => {
  if(redirectParams) {
    return decodeURIComponent(redirectParams)
  }
  return fallback;
}

const getAuthenticatedRoutes = () => {
  return mount({
    [routeRegistry.verifyUserInfo.path]: route(async (req, context) => {
      const { authService } = context;
      return {
        title: "Verify your information",
        view: <VerifyUserInfoFormComponent authService={authService} />
      }
    }),
    [routeRegistry.registerEntity.path]: getRegisterRoute(),
    [routeRegistry.dashboard.path]: map(async (request,context) => {
      const { authService } = context;
      if(!authService.currentUser.name) return redirect(routeRegistry.verifyUserInfo.path);
      if(!authService.currentEntity || !authService.currentEntity.verificationTxn) return redirect(routeRegistry.registerEntity.path);
      return route({
        title: "Diwala",
        view: <Dashboard currentEntity={authService.currentEntity}/>,
      })
    }),
    [routeRegistry.courses.path]: getCoursesRoute(),
    [routeRegistry.organization.path]: route({
      title: 'Organization',
      view: <Organization />
    }),
    [routeRegistry.userSettings.path]: route((request, context)=>{
      const { authService } = context;
      return{
        title: 'Your settings',
        view: <UserSettings authService={authService} />
      }
    })
  })
}


const getRegisterRoute = () => {
  return mount({
    '/': map(async(req, context) => {
      const { authService } = context;
      if(_.isEmpty(authService.currentEntity)) {
        return route({
          title: "Register your entity",
          view: <RegisterEntity authService={authService} />
        })
      }
      return redirect('/register/unverified');
    }),
    '/unverified': route(async(req, context) => {
      const { authService: { currentUser }} = context;
      return {
        title: 'Your entity has to be verified',
        view: <UnverifiedEntity currentUser={currentUser} />
      }
    }),

  })
}
```

**Comment by jamesknelson** [comment]
One possible reason that this is happening is that if `authService.loading` becomes true *after* the routes have already loaded, e.g. when login is required, then this line here will unmount the router:

```js
if(authService.loading) return <LoadLayout authService={authService} />
```

You can check if this is the case by putting a `console.log` with this line and checking if it is appearing in the console when the issue occurs.

Other than that, does `<AuthService.Provider>` always render its children? Or does it occassionally render nothing?

**Comment by vongohren** [comment]
@jamesknelson thanks for tips, I will check it out within the next couple of days I believe
The authservice provider comes from unstated hooks: https://github.com/jamiebuilds/unstated-next
So no rendering happening there.

**Comment by vongohren** [comment]
Well, what happened now was that I did not need that navigate there after all.
Because the auth service will login at that point, and the authservice set state function will make sure the router activates again and redirects/renders to the new right route

**Comment by strass** [comment]
I'm hitting a similar error, except `this` itself (of Navigation.prototype.navigate) is undefined. My Router is the top-level component and is not being un/remounted. Do you have any clue why this could be?

**Comment by strass** [comment]
I replicated the error here: https://codesandbox.io/s/dreamy-platform-c6b5g

**Comment by jamesknelson** [comment]
@strass the issue here is that `navigate` isn't bound Navigation class, so you can't destructure it -- you'll need to call it as a method:

```
navigation.navigate(...)
```

Here's a codesandbox with a fix: https://codesandbox.io/s/infallible-rosalind-cskfu

If you want, I'd be happy to submit a PR that binds it so that it is possible to destructure.

**Comment by strass** [comment]
Aaah, that explains a lot. I think that it should be possible to destructure or the typescript defs should be updated so that this case errors.

**Comment by jamesknelson** [comment]
Would be happy to update TS defs, but is it even possible to type it that way? I'm not familiar with the right approach.

**Comment by strass** [comment]
I'm not sure either – I think this is an artifact of typescript's duck typing, where it doesn't care whether an instance is a class or an object as long as it has all the right properties. I skimmed the docs and none of the class methods seem to cover this... Might be easier to just make it destructurable ;)

**Comment by jamesknelson** [comment]
Yeah. I'm not sure when I'll have time, but would definitely accept a PR! Should just be a matter of binding the `navigate` functions.

**Comment by jamesknelson** [comment]
Fixed in 0.13.6
